core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -2242,70 +2252,110 @@ private[spark] class DAGScheduler(
+              shuffleSendFinalizeRpcExecutor.submit(new Runnable() {
+                override def run(): Unit = {
+                  shuffleClient.finalizeShuffleMerge(shuffleServiceLoc.host,
+                    shuffleServiceLoc.port, shuffleId, shuffleMergeId,
+                    new MergeFinalizerListener {
+                      override def onShuffleMergeSuccess(statuses: MergeStatuses): Unit = {
+                        assert(shuffleId == statuses.shuffleId)
+                        eventProcessLoop.post(RegisterMergeStatuses(stage, MergeStatus.
+                          convertMergeStatusesToMergeStatusArr(statuses, shuffleServiceLoc)))
+                        results(index).set(true)
+                      }
 
-                override def onShuffleMergeFailure(e: Throwable): Unit = {
-                  logWarning(s"Exception encountered when trying to finalize shuffle " +
-                    s"merge on ${shuffleServiceLoc.host} for shuffle $shuffleId", e)
-                  // Do not fail the future as this would cause dag scheduler to prematurely
-                  // give up on waiting for merge results from the remaining shuffle services
-                  // if one fails
-                  results(index).set(false)
+                      override def onShuffleMergeFailure(e: Throwable): Unit = {
+                        logWarning(s"Exception encountered when trying to finalize shuffle " +
+                          s"merge on ${shuffleServiceLoc.host} for shuffle $shuffleId", e)
+                        // Do not fail the future as this would cause dag scheduler to prematurely
+                        // give up on waiting for merge results from the remaining shuffle services
+                        // if one fails
+                        if (e.isInstanceOf[IOException]) {
+                          logInfo(s"Failed to connect external shuffle service " +
+                            s"${shuffleServiceLoc.hostPort}")
+                          blockManagerMaster.removeShufflePushMergerLocation(shuffleServiceLoc.host)
+                          results(index).set(false)
+                        }
+                      }
+                    })
                 }
               })
+          }
         }
-      }
-      // DAGScheduler only waits for a limited amount of time for the merge results.
-      // It will attempt to submit the next stage(s) irrespective of whether merge results
-      // from all shuffle services are received or not.
+      var timedOut = false
       try {
         Futures.allAsList(results: _*).get(shuffleMergeResultsTimeoutSec, TimeUnit.SECONDS)
       } catch {
         case _: TimeoutException =>
+          timedOut = true
           logInfo(s"Timed out on waiting for merge results from all " +
             s"$numMergers mergers for shuffle $shuffleId")
       } finally {
+        if (timedOut || !registerMergeResults) {
+          cancelFinalizeShuffleMergeFutures(scheduledFutures,
```

Comment:
I don't think we should do this. Do we get any benefits from cancelling any pending send RPC tasks? I understand that when the timeout is elapsed, then the driver will not use the results from the shuffle service for that shuffle but at least the shuffle service can process the finalize request and close all the files for that shuffle. 
At the least, I think we shouldn't be clubbing this behavior change with the bug fix.
I know there was a discussion about this and sorry for commenting late on this.
cc @mridulm

Reply:
When `CONNECTIONCREATIONTIMEOUT * ((Lost_Node_Number -1) / Send_Finalize_Rpc_Thread_pool_size + 1) > shuffleMergeResultsTimeoutSec`, those futures will block the other send finalize tasks for the other stages. And shuffleMergeResultsTimeoutSec is short for small shuffle job. So I think this is helpful.

WDYT?

Reply:
Hmmm. Even if the send finalize tasks of future stages get blocked, with the other changes here, DAGScheduler will not stall and will schedule next stages. Just that the finalization of the future stages will just timeout. And since we are talking about small shuffle job, it is fine that push/merge didn't succeed. We are adding this code to address an extreme case but that increases the chances of shuffle service not cleaning up metadata/closing shuffle files which it does during finalization.
And ideally to address these issues the `connectionCreationTimeout` should be lowered. This was one of the reasons to introduce that configuration.

Reply:
You are right that the stage itself at driver wont stall after this PR @otterc - but if we dont cancel on timeout, a reduced number of mergers start receiving the finalization message in time, and we start seeing decreased merge ratio (or until the network situation improves).

Cancelling the task mitigates this issue.
The drawback being, if we had waited for some `X` seconds more, it is possible that the message could have made it through to ESS ... and so now we have a potential resource leak at shuffle service which is mitigated only when application exits.
Unfortunately, we dont have an upper bound of how long we can wait for restoring n/w operations (note - this is no ESS being down, that is immediately identified by a conn refusal - but rather n/w issue causing conn establishment or message send to block for an extended period of time)

Reply:
Yes, you are right. I would prefer a decreased merge ratio than increasing the chances of resource leak at the ESS. Reason being that push/merge is best effort. When it fails for an app, that single app is impacted. The app will just be slower. With issues on the server, multiple applications get impacted.
This is just my opinion. I am fine with whatever you folks decide.

Reply:
@wankunde, let me sync with @otterc offline - there are some nuances in here which I want to thrash out and make sure does not have issues.

Reply:
Discussed with @otterc - let us fix the current issue as discussed above.
I will let @otterc explore the problem of how to minimize potential for unclosed file handles due to the task cancellation, and due to transient failure to send merge finalize request in a separate jira.

Please fix the other comments in the PR @wankunde - thanks !

Reply:
Resolving comment thread